tensorflow/core/transforms/remapper/remapping_helper.h

```diff
@@ -0,0 +1,260 @@
+    }
+  }
+
+  // This function is currently used by convolution type op
+  bool IsCpuCompatibleDataFormat(Operation* conv_op,
+                                 const StringRef& attr_name = "data_format") {
+    StringRef data_format;
+    if (conv_op->hasAttrOfType<StringAttr>(attr_name)) {
+      data_format = conv_op->getAttrOfType<StringAttr>(attr_name).getValue();
+    } else {
+      return false;
+    }
+    if (dialect_->IsConv2D(conv_op)) {
+      return data_format == "NHWC" ||
+             (is_onednn_enabled_ && data_format == "NCHW");
+    } else if (dialect_->IsConv3D(conv_op)) {
+      return data_format == "NDHWC" ||
+             (is_onednn_enabled_ && data_format == "NCDHW");
+    } else {
+      return false;
+    }
+  }
+
+  bool IsGpuCompatible(const ContractionBiasAddActivation& pattern) {
+#if TENSORFLOW_USE_ROCM
+    // ROCm does not support _FusedConv2D. Does it suppport _FusedMatMul?
+    return false;
+#endif
+    // The TF->XLA bridge does not support `_FusedMatMul` so we avoid creating
+    // this op. Furthermore, XLA already does this fusion internally so there
+    // is no true benefit from doing this optimization if XLA is going to
+    // compile the unfused operations anyway.
+    if (xla_auto_clustering_) return false;
+    if (!util::NodeIsOnGpu(pattern.contraction)) return false;
+    if (!dialect_->IsRelu(pattern.activation)) return false;
+    if (dialect_->IsMatMul(pattern.contraction)) {
+      return IsGpuCompatibleDataType(pattern.contraction);
+    } else {
+      // TODO(intel-tf): Add spatial convolution support on GPU
+      return false;
+    }
+  }
+
+  // Currently GPU does not supprt contraction + bias_add
+  bool IsGpuCompatible(const ContractionBiasAdd&) { return false; }
+
+  bool IsCpuCompatible(Operation* contraction_op) {
+    if (!util::NodeIsOnCpu(contraction_op)) return false;
+    if (dialect_->IsConv2D(contraction_op) ||
+        dialect_->IsConv3D(contraction_op)) {
+      return IsCpuCompatibleDataType(contraction_op) &&
+             IsCpuCompatibleDataFormat(contraction_op);
+    } else if (dialect_->IsMatMul(contraction_op) ||
+               dialect_->IsAnyBatchMatMul(contraction_op) ||
+               dialect_->IsDepthwiseConv2dNative(contraction_op)) {
+      return IsCpuCompatibleDataType(contraction_op);
+    } else {
+      return false;
+    }
+  }
+
+  template <typename Pattern>
+  bool IsDeviceCompatible(const Pattern& pattern) {
+    // Currently, this function is used by contraction based fussion.
+    if constexpr (!std::is_same<Pattern, ContractionBiasAdd>::value &&
+                  !std::is_same<Pattern, ContractionBiasAddActivation>::value &&
+                  !std::is_same<Pattern, ContractionBiasAddAdd>::value &&
+                  !std::is_same<Pattern, ContractionBiasAddActivation>::value) {
+      return false;
+    }
+    return IsGpuCompatible(pattern) || IsCpuCompatible(pattern.contraction);
+  }
+
+ private:
+  TFGraphDialect* dialect_;
```

Comment:
This should be usable from the base class (make it protected there if it isn't already?)